golem-worker-service-base/src/service/api_definition_validator.rs

```diff
@@ -0,0 +1,158 @@
+use std::{collections::HashMap, sync::Arc};
+
+use crate::{
+    api_definition::{ApiDefinition, MethodPattern, PathPattern, Route},
+    service::template_service::TemplateService,
+};
+use async_trait::async_trait;
+use futures::future::join_all;
+use golem_common::model::TemplateId;
+use golem_service_base::model::{Export, ExportFunction, ExportInstance, Template};
+use serde::{Deserialize, Serialize};
+
+#[async_trait]
+pub trait ApiDefinitionValidatorService {
+    async fn validate(&self, api: &ApiDefinition) -> Result<(), ValidationError>;
+}
+
+#[derive(Debug, Clone, PartialEq, Default, Serialize, Deserialize, thiserror::Error)]
+// TODO: Fix this display impl.
+#[error("Validation error: {errors:?}")]
+pub struct ValidationError {
+    pub errors: Vec<RouteValidationError>,
+}
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RouteValidationError {
+    pub method: MethodPattern,
+    pub path: PathPattern,
+    pub template: TemplateId,
+    pub detail: String,
+}
+
+impl RouteValidationError {
+    fn from_route(route: Route, detail: String) -> Self {
+        Self {
+            method: route.method,
+            path: route.path,
+            template: route.binding.template,
+            detail,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct ApiDefinitionValidatorDefault {
+    template_service: Arc<dyn TemplateService + Send + Sync>,
+}
+
+impl ApiDefinitionValidatorDefault {
+    pub fn new(template_service: Arc<dyn TemplateService + Send + Sync>) -> Self {
+        Self { template_service }
+    }
+}
+
+#[async_trait]
+impl ApiDefinitionValidatorService for ApiDefinitionValidatorDefault {
+    async fn validate(&self, api: &ApiDefinition) -> Result<(), ValidationError> {
+        let get_templates = api
+            .routes
+            .clone()
+            .into_iter()
+            .map(|route| (route.binding.template.clone(), route))
+            .collect::<HashMap<_, _>>()
+            .into_values()
+            .map(|route| async move {
+                let id = &route.binding.template;
+                // TODO: Better error message.
+                self.template_service.get_latest(id).await.map_err(|e| {
+                    RouteValidationError::from_route(
+                        route,
+                        format!("Error getting latest template: {:?}", e),
```

Comment:
i do not think is good idea to have full error there, internal errors should not leek there, 

i think for beginning we should just "Error getting latest template", later we can do it better

Reply:
yeah I put the `TODO` there haha. I will remove the details and will revisit later